Add AWS hardware MFA matcher

pwelch opened the pull request:
Adding a hardware as well as a virtual MFA matcher for aws_iam_root_user
resource

Fixes #2750

Signed-off-by: Paul Welch <pwelch@chef.io>

Commit:
Add AWS hardware MFA matcher

Adding a hardware as well as a virtual MFA matcher for aws_iam_root_user
resource

Fixes #2750

Signed-off-by: Paul Welch <pwelch@chef.io>

clintoncwolfe commented on test/unit/resources/aws_iam_root_user_test.rb:
```diff
@@ -44,4 +44,52 @@ def test_has_mfa_enabled_returns_false_when_account_mfa_devices_is_zero
   end
-end
+
+  def test_has_virtual_mfa_enabled_returns_true_when_account_vmfa_devices_is_one
```
:100: for meaningful unit test names

clintoncwolfe commented on lib/resources/aws/aws_iam_root_user.rb:
```diff
@@ -46,6 +46,20 @@ def has_mfa_enabled?
     summary_account['AccountMFAEnabled'] == 1
   end
 
+  def has_virtual_mfa_devices?
```
So, it turns out that even though the API returns an array, you can only have one MFA device total per root account or IAM user (see [FAQ](https://aws.amazon.com/iam/faqs/#MFA_FAQs), 'Q. Can I have multiple authentication devices active for my AWS account?').  

Based on that, I think we should rename this matcher to be `has_virtual_mfa_enabled?`.

pwelch replied:
TIL!

clintoncwolfe commented on lib/resources/aws/aws_iam_root_user.rb:
```diff
@@ -46,6 +46,20 @@ def has_mfa_enabled?
+    false
+  end
+
+  def has_hardware_mfa_devices?
```
Likewise, this should be renamed `has_hardware_mfa_enabled?`.

clintoncwolfe commented on lib/resources/aws/aws_iam_root_user.rb:
```diff
@@ -46,6 +46,20 @@ def has_mfa_enabled?
 
+  def has_virtual_mfa_devices?
+    virtual_mfa_devices.each do |device|
+      if %r{arn:aws:iam::\d{12}:mfa\/root-account-mfa-device} =~
```
Might want to add a comment here - if the root account has a VMFA, it will have this special serial number, ending in 'root-account-mfa-device'.

clintoncwolfe commented on lib/resources/aws/aws_iam_root_user.rb:
```diff
@@ -57,4 +71,8 @@ def summary_account
   end
+
+  def virtual_mfa_devices
+    @__virtual_devices ||= @client.list_virtual_mfa_devices.virtual_mfa_devices
```
Need to wrap this in a `catch_aws_errors` block.

pwelch replied:
👍

clintoncwolfe commented on lib/resources/aws/aws_iam_root_user.rb:
```diff
@@ -46,6 +46,20 @@ def has_mfa_enabled?
+  end
+
+  def has_hardware_mfa_devices?
+    @__hardware_devices ||= !@client.list_mfa_devices.mfa_devices.empty?
```
Alas, this isn't the right logic.  `list_mfa_devices` [will list mfa devices for a particular user](https://docs.aws.amazon.com/sdkforruby/api/Aws/IAM/Client.html#list_mfa_devices-instance_method), or the _currently authenticated user_ - which would be the test harness user, not the root user.  You can't pass the username 'root', either:
```
[cwolfe@lodi inspec-review]$ aws --profile inspec-aws-test-default iam list-mfa-devices --user-name root

An error occurred (NoSuchEntity) when calling the ListMFADevices operation: The user with name root cannot be found.
```

So that's no fun.  However, the CIS PDF tells us that if the account has MFA enabled (which we know from `get_account_summary`, and we have a predicate for it in `has_mfa_enabled?`) BUT has no virtual MFA device (which you have added a predicate for), then it must have a hardware MFA.  So I think this can simply be:

```
   has_mfa_enabled? && !has_virtual_mfa_enabled?
```

clintoncwolfe commented on test/unit/resources/aws_iam_root_user_test.rb:
```diff
@@ -44,4 +44,52 @@ def test_has_mfa_enabled_returns_false_when_account_mfa_devices_is_zero
+
+    assert_equal false, AwsIamRootUser.new(@mock_conn).has_hardware_mfa_devices?
+  end
+end
```
One missing piece: you need to add the two new matchers to the docs, at `docs/resources/aws_iam_root_account.md.erb`.  Thanks!

clintoncwolfe reviewed (changes requested):
A great start!  I suggest a couple of renames, and I found a logic error.  And we need docs!  Great unit tests, though - this is coming along nicely!

Commit:
Add New AWS Root Matcher Docs

- Add documentation for new root MFA matchers
- Fix logic for checking MFA devices from feedback on PR

Signed-off-by: Paul Welch <pwelch@chef.io>

TrevorBramble commented on lib/resources/aws/aws_iam_root_user.rb:
```diff
@@ -46,6 +46,22 @@ def has_mfa_enabled?
+      end
+    end
+    false
+  end
```
How about extracting the pattern to an explaining variable and using `Enumerable#any?()` instead?
```ruby
# if the root account has a Virtual MFA device then it will have a special
# serial number ending in 'root-account-mfa-device'
def has_virtual_mfa_enabled?
  mfa_device_pattern = %r{arn:aws:iam::\d{12}:mfa\/root-account-mfa-device}

  virtual_mfa_devices.any? { |device| mfa_device_pattern =~ device['serial_number'] }
end
```

TrevorBramble commented on test/unit/resources/aws_iam_root_user_test.rb:
```diff
@@ -44,4 +44,54 @@ def test_has_mfa_enabled_returns_false_when_account_mfa_devices_is_zero
+        serial_number: "arn:aws:iam::123456789011:mfa/root-account-mfa-device",
+        user: Aws::IAM::Types::User.new(
+          user_id: "123456789011",
+          arn:     "arn:aws:iam::123456789011:root",
```
I think we're aligning single quotes in this project?

Commit:
Add Integration tests for MFA matchers

- Add integration tests for virtual and hardware MFA matchers
- Clean up logic for has_virtual_mfa_enabled? method

Signed-off-by: Paul Welch <pwelch@chef.io>

jquick reviewed (approved):
Thanks @pwelch !
